### ISSUE TITLE
Is there a way to modify the templates / HTML?

### ISSUE THREAD

**Issue by jameshoward** [rtfpessoa/diff2html#100]
We are using your awesome project to display a diff before importing files into a webapp. What we're hoping to achieve is having a checkbox next to each filename title to select whether to include that file in the import.

The first thought I had was to add checkboxes with JS after but it would be useful to have more details on the file and build the HTML at once. I noticed there are templates built with Hogan JS but I don't see how they could be changed/overridden without forking this project and building our own version.

Any thoughts would be great, thanks.

**Comment by rtfpessoa** [comment]
Currently there is no easy way to make this happen out of the box.

That use case was already mentioned several times and is definitely I want to have.

Just to start the discussion. There are several ways this can be done:

* Allow to inject templates.

> This is probably the faster way from the tool side but the user needs to provide full html for a part. If the parts are small enough this might not be that bad. 

* Have some kind of subscription that you can hook to. After one html element is generated you receive the element and can apply some changes and return the new element.

> This is harder and I am not sure if there is an easy way for you to manipulate the element.


Summing it up, I guess the first one would be better, would you be interested in providing hogan.js templates? Maybe we could even accept compiled and not compiled templates.

Any suggestions?

**Comment by jameshoward** [comment]
Not saying this is the 'right' way but what I've seen elsewhere is a way to override a template like your option 1. So something like this in the config object:

```
{
    templates: {
        'file-summary-wrapper': …
    }
}
```

Then any that aren't overridden use the default, library-provided versions.

I've not used Hogan so I'm not familiar with it but I'm getting the `…` above could be like the existing templates, e.g. `new Hogan.Template(…)`.

**Comment by rtfpessoa** [comment]
Any pull requests are welcome. I can try to play with this as soon as I get some time.

**Comment by rtfpessoa** [comment]
@jameshoward today I got some time to play around with this.

I made a pretty simple solution in #106 that provides a way to override templates through the initial configuration passed to diff2html.

Let me know if it works for your use case.

**Comment by jameshoward** [comment]
It looks good to me. I'll be working on the same piece again tomorrow so will check it works but looks promising.

**Comment by jameshoward** [comment]
Checked out  that branch and got running with it. I suppose the barrier to entry is having to compile the Hogan template and pass it in when the library (quite rightly) doesn't expose Hogan or the JS utils.

It works fine for my case though, where I can pre-compile the template and use Hogan directly so I say merge and release! 🎉

**Comment by rtfpessoa** [comment]
@jameshoward I exposed one method to help compilation in HoganJsUtils, do you think I should expose it in the main API of the library?

**Comment by jameshoward** [comment]
I'm using the library in a browser so I'm assuming the only globals are `Diff2Html` and `Diff2HtmlUI` and there's no way to compile a template through either?

I'm not sure about exposing it on the API. There would be pros and cons, I'm sure.

**Comment by rtfpessoa** [comment]
I can accept uncompiled templates, maybe it helps.

Initially I thought it would not be interesting, since people might want to cache it.

**Comment by rtfpessoa** [comment]
This was released as version 2.0.8

**Comment by jameshoward** [comment]
That's really cool, thank you. I've used the 2.0.8 version to do exactly what I need.

**Comment by brian-lagerman** [comment]
I'm going to hang my question here cause it's very related. I've managed to supply the rawTemplates in their uncompiled versions, and that's working well. Now I'm trying to compile them, so I cloned the repo, changed my templates, and ran the 'templates' script which outputs chunks to 'diff2html-templates.js' 'like:

`
global.browserTemplates["generic-line"] = new Hogan.Template({code: function (c,p,i) {  <more js...> ; },partials: {}, subs: {  }});
`

My question is, am I done? i.e., within `getPrettyHtml()` just pass this line into "templates" like so:

`
templates: {
  'generic-line': new Hogan.Template({code: function (c,p,i) { etc, etc, }})
 }
`

Or do I have another utility to run?
